include/revng/Pipeline/ReadPathsCache.h

```diff
@@ -0,0 +1,215 @@
+
+class ParsableLocatedTarget {
+public:
+  std::string Content;
+  std::string StepName;
+  std::string ContainerName;
+};
+
+class LocatedTarget {
+private:
+  Target Content;
+  std::string StepName;
+  std::string ContainerName;
+
+public:
+  LocatedTarget(const Target &Target,
+                const std::string &StepName,
+                const std::string &ContainerName) :
+    Content(Target), StepName(StepName), ContainerName(ContainerName) {}
+
+  static llvm::Expected<llvm::SmallVector<LocatedTarget, 2>>
+  fromParsable(const Context &Ctx, const ParsableLocatedTarget &Target);
+
+  const Target &getTarget() const { return Content; }
+
+  llvm::StringRef getStepName() const { return llvm::StringRef(StepName); }
+
+  llvm::StringRef getContainerName() const {
+    return llvm::StringRef(ContainerName);
+  }
+
+  Target &getTarget() { return Content; }
+
+  std::string &getStepName() { return StepName; }
+
+  std::string &getContainerName() { return ContainerName; }
+
+  bool operator==(const LocatedTarget &) const = default;
+
+  bool operator!=(const LocatedTarget &) const = default;
+
+  bool operator<(const LocatedTarget &Other) const {
+    const auto &LHS = std::tie(Content, StepName, ContainerName);
+    return LHS < std::tie(Other.Content, Other.StepName, Other.ContainerName);
+  }
+
+  ParsableLocatedTarget toParsable() const;
+};
+
+class TargetReadPathsSerializable {
+public:
+  std::string Path;
+};
+
+class PathSerializationPrimitives {
+public:
+  PathSerializationPrimitives() = default;
+
+  template<TupleTreeCompatibleAndVerifiable T>
+  static PathSerializationPrimitives make() {
+    PathSerializationPrimitives Primitives;
+    Primitives.Deserialize = &stringAsPath<T>;
+    Primitives.Serialize = &pathAsString<T>;
+    return Primitives;
+  }
+
+  std::optional<TupleTreePath> (*Deserialize)(llvm::StringRef);
+  std::optional<std::string> (*Serialize)(const TupleTreePath &);
+};
+
+class TargetReadPaths {
+private:
+  TupleTreePath Path;
+
+public:
+  explicit TargetReadPaths(const TupleTreePath &Path) : Path(Path) {}
+
+  static llvm::Expected<TargetReadPaths>
+  fromParsable(const Context &Ctx,
+               const TargetReadPathsSerializable &Target,
+               PathSerializationPrimitives Primitives);
+
+  const TupleTreePath &getPaths() const { return Path; }
+
+  TupleTreePath &getPaths() { return Path; }
+
+  bool operator==(const TargetReadPaths &Other) const = default;
+  bool operator!=(const TargetReadPaths &Other) const = default;
+
+  bool operator<(const TargetReadPaths &Other) const {
+    return std::tie(Path) < std::tie(Other.Path);
+  }
+
+  TargetReadPathsSerializable
+  toParsable(PathSerializationPrimitives Primitives) const;
+};
+
+class SerializableReadPathCache {
+public:
+  using ValueType = std::pair<pipeline::TargetReadPathsSerializable,
+                              pipeline::ParsableLocatedTarget>;
+  using Vector = std::vector<ValueType>;
+  Vector Data;
+};
+
+class ReadPathCache {
```

Comment:
This is a bit of a loaded name that implies its usages, not what it does "in the void".

Maybe we can call this `PathToTargetBimap` or something along that line?